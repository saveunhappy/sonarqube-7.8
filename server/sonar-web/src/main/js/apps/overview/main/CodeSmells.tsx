/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import enhance, { ComposedProps } from './enhance';
import DocTooltip from '../../../components/docs/DocTooltip';
import { getMetricName } from '../utils';
import { formatMeasure } from '../../../helpers/measures';
import CodeSmellIcon from '../../../components/icons-components/CodeSmellIcon';
import DrilldownLink from '../../../components/shared/DrilldownLink';

export class CodeSmells extends React.PureComponent<ComposedProps> {
  renderHeader() {
    return this.props.renderHeader('Maintainability');
  }

  renderDebt(metric: string) {
    const { branchLike, measures, component } = this.props;
    const measure = measures.find(measure => measure.metric.key === metric);
    const value = measure ? this.props.getValue(measure) : undefined;

    return (
      <DrilldownLink branchLike={branchLike} component={component.key} metric={metric}>
        {formatMeasure(value, 'SHORT_WORK_DUR')}
      </DrilldownLink>
    );
  }

  renderTimeline(range: string) {
    return this.props.renderTimeline('sqale_index', range);
  }

  renderLeak() {
    const { leakPeriod } = this.props;
    if (!leakPeriod) {
      return null;
    }

    return (
      <div className="overview-domain-leak">
        <div className="overview-domain-measures">
          <div className="overview-domain-measure">
            <div className="overview-domain-measure-value">
              <span className="offset-left">{this.renderDebt('new_technical_debt')}</span>
              {this.props.renderRating('new_maintainability_rating')}
            </div>
            <div className="overview-domain-measure-label">{getMetricName('new_effort')}</div>
          </div>
          <div className="overview-domain-measure">
            <div className="overview-domain-measure-value">
              {this.props.renderIssues('new_code_smells', 'CODE_SMELL')}
            </div>
            <div className="overview-domain-measure-label">
              <CodeSmellIcon className="little-spacer-right" />
              {getMetricName('new_code_smells')}
            </div>
          </div>
        </div>
        {this.renderTimeline('after')}
      </div>
    );
  }

  renderNutshell() {
    return (
      <div className="overview-domain-nutshell">
        <div className="overview-domain-measures">
          <div className="overview-domain-measure">
            <div className="overview-domain-measure-value">
              <span className="offset-left">{this.renderDebt('sqale_index')}</span>
              {this.props.renderRating('sqale_rating')}
            </div>
            <div className="overview-domain-measure-label display-flex-center display-flex-justify-center">
              {getMetricName('effort')}
              <DocTooltip
                className="little-spacer-left"
                doc={import(/* webpackMode: "eager" */ 'Docs/tooltips/metrics/debt.md')}
              />
            </div>
            {this.props.renderHistoryLink('sqale_index')}
          </div>
          <div className="overview-domain-measure">
            <div className="overview-domain-measure-value">
              {this.props.renderIssues('code_smells', 'CODE_SMELL')}
            </div>
            <div className="overview-domain-measure-label display-flex-center display-flex-justify-center">
              <CodeSmellIcon className="little-spacer-right " />
              {getMetricName('code_smells')}
              <DocTooltip
                className="little-spacer-left"
                doc={import(/* webpackMode: "eager" */ 'Docs/tooltips/metrics/code-smells.md')}
              />
            </div>
            {this.props.renderHistoryLink('code_smells')}
          </div>
        </div>
        {this.renderTimeline('before')}
      </div>
    );
  }

  render() {
    const { measures } = this.props;
    const codeSmellsMeasure = measures.find(measure => measure.metric.key === 'code_smells');
    if (!codeSmellsMeasure) {
      return null;
    }
    return (
      <div className="overview-card" id="overview-code-smells">
        {this.renderHeader()}

        <div className="overview-domain-panel">
          {this.renderNutshell()}
          {this.renderLeak()}
        </div>
      </div>
    );
  }
}

export default enhance(CodeSmells);
