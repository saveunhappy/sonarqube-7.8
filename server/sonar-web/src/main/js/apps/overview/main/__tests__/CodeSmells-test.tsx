/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { shallow } from 'enzyme';
import CodeSmells from '../CodeSmells';
import { ComposedProps } from '../enhance';
import {
  mockComponent,
  mockMainBranch,
  mockMeasureEnhanced,
  mockMetric
} from '../../../../helpers/testMocks';

it('should render correctly', () => {
  expect(shallowRender()).toMatchSnapshot();
});

function shallowRender(props: Partial<ComposedProps> = {}) {
  return shallow(
    <CodeSmells
      branchLike={mockMainBranch()}
      component={mockComponent()}
      history={{ sqale_index: [] }}
      leakPeriod={{ index: 1 } as T.Period}
      measures={[
        mockMeasureEnhanced({
          metric: mockMetric({
            id: 'code_smells',
            key: 'code_smells',
            name: 'Code Smells',
            type: 'INT'
          }),
          value: '15'
        }),
        mockMeasureEnhanced({
          metric: mockMetric({
            id: 'sqale_index',
            key: 'sqale_index',
            name: 'Debt',
            type: 'INT'
          }),
          value: '1052'
        }),
        mockMeasureEnhanced({
          metric: mockMetric({
            id: 'sqale_rating',
            key: 'sqale_rating',
            name: 'Maintainability',
            type: 'RATING'
          }),
          value: '2.0'
        }),
        mockMeasureEnhanced({
          metric: mockMetric({
            id: 'new_code_smells',
            key: 'new_code_smells',
            name: 'New Code Smells',
            type: 'INT'
          }),
          periods: [
            {
              bestValue: true,
              index: 1,
              value: '52'
            }
          ]
        }),
        mockMeasureEnhanced({
          metric: mockMetric({
            id: 'new_technical_debt',
            key: 'new_technical_debt',
            name: 'New Debt',
            type: 'INT'
          }),
          periods: [
            {
              bestValue: true,
              index: 1,
              value: '85'
            }
          ]
        }),
        mockMeasureEnhanced({
          metric: mockMetric({
            id: 'new_maintainability_rating',
            key: 'new_maintainability_rating',
            name: 'New Maintainability',
            type: 'RATING'
          }),
          value: '3.0'
        })
      ]}
      {...props}
    />
  ).dive();
}
