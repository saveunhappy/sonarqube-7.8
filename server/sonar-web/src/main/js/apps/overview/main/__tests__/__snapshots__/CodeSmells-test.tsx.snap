// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div
  className="overview-card"
  id="overview-code-smells"
>
  <div
    className="overview-card-header"
  >
    <div
      className="overview-title"
    >
      <span>
        metric_domain.Maintainability
      </span>
      <Link
        className="spacer-left small"
        onlyActiveOnIndex={false}
        style={Object {}}
        to={
          Object {
            "pathname": "/component_measures",
            "query": Object {
              "id": "my-project",
              "metric": "Maintainability",
            },
          }
        }
      >
        layout.measures
      </Link>
    </div>
  </div>
  <div
    className="overview-domain-panel"
  >
    <div
      className="overview-domain-nutshell"
    >
      <div
        className="overview-domain-measures"
      >
        <div
          className="overview-domain-measure"
        >
          <div
            className="overview-domain-measure-value"
          >
            <span
              className="offset-left"
            >
              <DrilldownLink
                branchLike={
                  Object {
                    "analysisDate": "2018-01-01",
                    "isMain": true,
                    "name": "master",
                  }
                }
                component="my-project"
                metric="sqale_index"
              >
                work_duration.x_days.2
              </DrilldownLink>
            </span>
            <Tooltip
              overlay="metric.sqale_rating.tooltip.B.0.0%"
            >
              <div
                className="overview-domain-measure-sup"
              >
                <DrilldownLink
                  branchLike={
                    Object {
                      "analysisDate": "2018-01-01",
                      "isMain": true,
                      "name": "master",
                    }
                  }
                  className="link-no-underline"
                  component="my-project"
                  metric="sqale_rating"
                >
                  <Rating
                    value="2.0"
                  />
                </DrilldownLink>
              </div>
            </Tooltip>
          </div>
          <div
            className="overview-domain-measure-label display-flex-center display-flex-justify-center"
          >
            overview.metric.effort
            <DocTooltip
              className="little-spacer-left"
              doc={Promise {}}
            />
          </div>
          <Link
            className="overview-domain-measure-history-link"
            onlyActiveOnIndex={false}
            style={Object {}}
            to={
              Object {
                "pathname": "/project/activity",
                "query": Object {
                  "custom_metrics": "sqale_index",
                  "graph": "custom",
                  "id": "my-project",
                },
              }
            }
          >
            <HistoryIcon />
            <span>
              project_activity.page
            </span>
          </Link>
        </div>
        <div
          className="overview-domain-measure"
        >
          <div
            className="overview-domain-measure-value"
          >
            <Link
              onlyActiveOnIndex={false}
              style={Object {}}
              to={
                Object {
                  "pathname": "/project/issues",
                  "query": Object {
                    "id": "my-project",
                    "resolved": "false",
                    "types": "CODE_SMELL",
                  },
                }
              }
            >
              15
            </Link>
          </div>
          <div
            className="overview-domain-measure-label display-flex-center display-flex-justify-center"
          >
            <CodeSmellIcon
              className="little-spacer-right "
            />
            overview.metric.code_smells
            <DocTooltip
              className="little-spacer-left"
              doc={Promise {}}
            />
          </div>
          <Link
            className="overview-domain-measure-history-link"
            onlyActiveOnIndex={false}
            style={Object {}}
            to={
              Object {
                "pathname": "/project/activity",
                "query": Object {
                  "custom_metrics": "code_smells",
                  "graph": "custom",
                  "id": "my-project",
                },
              }
            }
          >
            <HistoryIcon />
            <span>
              project_activity.page
            </span>
          </Link>
        </div>
      </div>
      <div
        className="overview-domain-timeline"
      >
        <Timeline
          history={Array []}
        />
      </div>
    </div>
    <div
      className="overview-domain-leak"
    >
      <div
        className="overview-domain-measures"
      >
        <div
          className="overview-domain-measure"
        >
          <div
            className="overview-domain-measure-value"
          >
            <span
              className="offset-left"
            >
              <DrilldownLink
                branchLike={
                  Object {
                    "analysisDate": "2018-01-01",
                    "isMain": true,
                    "name": "master",
                  }
                }
                component="my-project"
                metric="new_technical_debt"
              >
                work_duration.x_hours.1
              </DrilldownLink>
            </span>
            <Tooltip
              overlay="metric.sqale_rating.tooltip.A.0.0%"
            >
              <div
                className="overview-domain-measure-sup"
              >
                <DrilldownLink
                  branchLike={
                    Object {
                      "analysisDate": "2018-01-01",
                      "isMain": true,
                      "name": "master",
                    }
                  }
                  className="link-no-underline"
                  component="my-project"
                  metric="new_maintainability_rating"
                >
                  <Rating
                    value="1.0"
                  />
                </DrilldownLink>
              </div>
            </Tooltip>
          </div>
          <div
            className="overview-domain-measure-label"
          >
            overview.metric.new_effort
          </div>
        </div>
        <div
          className="overview-domain-measure"
        >
          <div
            className="overview-domain-measure-value"
          >
            <Link
              onlyActiveOnIndex={false}
              style={Object {}}
              to={
                Object {
                  "pathname": "/project/issues",
                  "query": Object {
                    "id": "my-project",
                    "resolved": "false",
                    "sinceLeakPeriod": "true",
                    "types": "CODE_SMELL",
                  },
                }
              }
            >
              52
            </Link>
          </div>
          <div
            className="overview-domain-measure-label"
          >
            <CodeSmellIcon
              className="little-spacer-right"
            />
            overview.metric.new_code_smells
          </div>
        </div>
      </div>
      <div
        className="overview-domain-timeline"
      >
        <Timeline
          history={Array []}
        />
      </div>
    </div>
  </div>
</div>
`;
